### PR TITLE
Feature/ml experiments

### PR DESCRIPTION
**Pull request by bangxiangyong**
- Implemented ML Experiments and agent pipelines to modularize and track the implementation and experimentation of multiple configurations of ML algorithms. 

- A second page in the dashboard is developed to specifically cater to viewing and comparing the ML experiments executed.

- Preliminary examples are provided to give a feel of what it can do.

### CODE REVIEW

**Comment by bangxiangyong** [comment]
Both #25 #17 are addressed in this branch

**Comment by bangxiangyong** [comment]
@BjoernLudwigPTB  perhaps you can run the examples in https://github.com/bangxiangyong/agentMET4FOF/tree/feature/ml_experiments/examples/ML_EXPERIMENTS 

to have a feel of it.

**Comment by BjoernLudwigPTB** [comment]
I Will now take a closer look on why the test coverage dropped so significantly from 
1728b82 to 31c85b8 and try to "fix" this to merge the PR. Looks great so far!

File: agentMET4FOF/dashboard/Dashboard_Control.py [hunk 1]
```diff
@@ -5,16 +5,19 @@
 
 
 #global variables access via 'dashboard_ctrl'
-class Dashboard_Control():
-    def __init__(self, ip_addr="127.0.0.1", port=3333, modules= []):
-        super(Dashboard_Control, self).__init__()
+class _Dashboard_Control():
```
Comment (BjoernLudwigPTB) [inline agentMET4FOF/dashboard/Dashboard_Control.py]: There is actually no such thing as a private class in a separate module according to PEP 8, as far as I know, although the underscore still clarifies, that we should not use this class for anything on its own. I think this should rather be refactored to be part of `Dashboard.py`.

File: agentMET4FOF/agents.py [hunk 2]
```diff
@@ -332,9 +340,22 @@ def bind_output(self, output_agent):
 
         Parameters
         ----------
-        output_agent : AgentMET4FOF
-            Agent to be binded to this agent's output channel
+        output_agent : AgentMET4FOF or list
+            Agent(s) to be binded to this agent's output channel
 
+        """
+        if "AgentPipeline" in str(type(output_agent).__name__):
```
[the remark this answers is not in the document]
Comment (BjoernLudwigPTB) [reply agentMET4FOF/agents.py]: I changed this myself in #42

File: agentMET4FOF/agents.py [hunk 3]
```diff
@@ -82,10 +85,11 @@ def on_init(self):
         self.log_mode = True
 
         self.output_channels_info = {}
+
         try:
             self.init_parameters()
-        except:
-            print("Error in calling init_parameters()...")
+        except Exception as e:
+            return 0
```
[the remark this answers is not in the document]
Comment (bangxiangyong) [reply agentMET4FOF/agents.py]: you are right, i must have made a mistake here, thanks for pointing out !

File: agentMET4FOF/agents.py [hunk 4]
```diff
@@ -332,9 +340,22 @@ def bind_output(self, output_agent):
 
         Parameters
         ----------
-        output_agent : AgentMET4FOF
-            Agent to be binded to this agent's output channel
+        output_agent : AgentMET4FOF or list
+            Agent(s) to be binded to this agent's output channel
```
[the remark this answers is not in the document]
Comment (bangxiangyong) [reply agentMET4FOF/agents.py]: English-wise yes, you're right on that ! ..

**Comment by BjoernLudwigPTB** [comment]
After the last change was reviewed we can finally merge this!

**Pull request merged**